### PR TITLE
service: caps: Implement SaveScreenShotEx0 and variants

### PR DESCRIPTION
**Pull request by german77**
Used by SSBU and probably more games. Allows to save screenshot captures from within games. Album screenshot editing is still not working.

01006a800016e000_2023-09-23_23-17-40-000

### CODE REVIEW

File: src/core/hle/service/caps/caps_manager.cpp [hunk 1]
```diff
@@ -365,6 +412,45 @@ Result AlbumManager::LoadImage(std::span<u8> out_image, const std::filesystem::p
...
+        fmt::format("{:04}-{:02}-{:02}_{:02}-{:02}-{:02}-{:03}", date.year, date.month, date.day,
+                    date.hour, date.minute, date.second, 0);
+    const std::string file_path =
+        fmt::format("{}/{:016x}_{}.png", screenshot_path, title_id, formatted_date);
+
+    const Common::FS::IOFile db_file{file_path, Common::FS::FileAccessMode::Write,
+                                     Common::FS::FileType::BinaryFile};
+
+    int len;
+    unsigned char* png = stbi_write_png_to_mem(image.data(), 0, 1280, 720, STBI_rgb_alpha, &len);
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_manager.cpp:429]: can this be a yuzu type instead?

File: src/core/hle/service/caps/caps_manager.cpp [hunk 2]
```diff
@@ -365,6 +412,45 @@ Result AlbumManager::LoadImage(std::span<u8> out_image, const std::filesystem::p
...
+
+    if (png == NULL) {
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_manager.cpp:431]:
```suggestion
    if (!png) {
```

File: src/core/hle/service/caps/caps_manager.cpp [hunk 3]
```diff
@@ -365,6 +412,45 @@ Result AlbumManager::LoadImage(std::span<u8> out_image, const std::filesystem::p
...
+        return ResultFileCountLimit;
+    }
+
+    std::vector<u8> png_image(len);
+    memcpy(png_image.data(), png, len);
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_manager.cpp:436]:
```suggestion
    std::memcpy(png_image.data(), png, len);
```

File: src/core/hle/service/caps/caps_manager.cpp [hunk 4]
```diff
@@ -365,6 +412,45 @@ Result AlbumManager::LoadImage(std::span<u8> out_image, const std::filesystem::p
...
+    const std::string formatted_date =
+        fmt::format("{:04}-{:02}-{:02}_{:02}-{:02}-{:02}-{:03}", date.year, date.month, date.day,
+                    date.hour, date.minute, date.second, 0);
+    const std::string file_path =
+        fmt::format("{}/{:016x}_{}.png", screenshot_path, title_id, formatted_date);
+
+    const Common::FS::IOFile db_file{file_path, Common::FS::FileAccessMode::Write,
+                                     Common::FS::FileType::BinaryFile};
+
+    int len;
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_manager.cpp:428]: dito

File: src/core/hle/service/caps/caps_su.cpp [hunk 5]
```diff
@@ -36,4 +39,62 @@ void IScreenShotApplicationService::SetShimLibraryVersion(HLERequestContext& ctx
...
+    const auto image_data_buffer = ctx.ReadBuffer(1);
+
+    LOG_INFO(Service_Capture,
+             "called, report_option={}, image_data_buffer_size={}, applet_resource_user_id={}",
+             parameters.report_option, image_data_buffer.size(),
+             parameters.applet_resource_user_id);
+
+    ApplicationAlbumEntry entry{};
+    ApplicationData app_data{};
+    memcpy(&app_data, app_data_buffer.data(), sizeof(ApplicationData));
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_su.cpp:90]:
```suggestion
    std::memcpy(&app_data, app_data_buffer.data(), sizeof(ApplicationData));
```

File: src/core/hle/service/caps/caps_ss.h [hunk 6]
```diff
@@ -13,8 +13,14 @@ namespace Service::Capture {
...
 public:
-    explicit IScreenShotService(Core::System& system_);
+    explicit IScreenShotService(Core::System& system_, std::shared_ptr<AlbumManager> album_manager);
     ~IScreenShotService() override;
+
+private:
+    void SaveScreenShotEx0(HLERequestContext& ctx);
+    void SaveEditedScreenShotEx1(HLERequestContext& ctx);
+
+    std::shared_ptr<AlbumManager> manager = nullptr;
```
Comment (FearlessTobi) [inline src/core/hle/service/caps/caps_ss.h:23]:
```suggestion
    std::shared_ptr<AlbumManager> manager;
```

seems like this always gets initialized in the constructor anyway.

**Review by FearlessTobi — APPROVED** [review]

**Comment by german77** [comment]
@abouvier I couldn't fix cmake to build without errors. So I ended up splitting the file again.

**Comment by abouvier** [comment]
> @abouvier I couldn't fix cmake to build without errors. So I ended up splitting the file again.

I will try once the PR is merged :p

**Comment by TonRZN** [comment]
Sorry to ask in a merged PR. But this has any relation to caps_u.cpp or the SaveCurrentScreenshot of the am.cpp? Games like Pokemon Scarlet/Violet uses those.

**Comment by german77** [comment]
I just tested scarlet. This game relies on you pressing the screenshot button on your joycon to actually take the picture. So this game doesn't use the caps service. You get the same effect by using the hotkey `ctrl+p` or pressing the screenshot button on your controller if you have that button mapped.

**Comment by TonRZN** [comment]
Oh, even on the new update? You can use the A button to capture too, didn't know about this quirk of the game.

**Comment by german77** [comment]
I didn't check with the update. Please include such details. That's definitely using caps service. I will try to implement it soon.

**Comment by german77** [comment]
Implemented here #11892

**Pull request merged**